README.md
# Acme::Ford::Prefect::FFI ![linux](https://github.com/PerlAlien/Acme-Ford-Prefect-FFI/workflows/linux/badge.svg)

FFI test module for Alien::Base

# SYNOPSIS

```perl
use Acme::Ford::Prefect::FFI;

my $answer = Acme::Ford::Prefect::FFI::answer(); # == 42 of course
```

# DESCRIPTION

[Alien::Base](https://metacpan.org/pod/Alien::Base) comprises base classes to help in the construction of `Alien::` modules.  Modules in the [Alien](https://metacpan.org/pod/Alien) namespace are used to locate and install (if necessary)
external libraries needed by other Perl modules.

This module is a toy module to test the efficacy of the [Alien::Base](https://metacpan.org/pod/Alien::Base) system with its experimental FFI interfaces.  This module depends on another toy module 
[Acme::Alien::DontPanic](https://metacpan.org/pod/Acme::Alien::DontPanic) which provides the needed libdontpanic library to be able to tell us the `answer` to life, the universe and everything.

# FUNCTIONS

## answer

```perl
my $answer = Acme::Ford::Prefect::FFI::answer();
```

Returns the answer to life the universe and everything.  Not exported.

# SEE ALSO

- [FFI::Platypus](https://metacpan.org/pod/FFI::Platypus)
- [Alien::Base](https://metacpan.org/pod/Alien::Base)
- [Alien](https://metacpan.org/pod/Alien)

# AUTHOR

Graham Ollis <plicease@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2014-2022 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Perl module {{$dist->name}}

{{$NEXT}}

2.7200    2022-10-26 06:32:04 -0600
  - Require Acme::Alien::DontPanic 2.7200 (Alien::Build 2.72)

2.6500    2022-08-30 11:25:19 -0600
  - Require Acme::Alien::DontPanic 2.6500 (Alien::Build 2.65)

2.6400    2022-08-30 10:21:21 -0600
  - Require Acme::Alien::DontPanic 2.6400 (Alien::Build 2.64)

2.5900    2022-08-16 09:53:01 -0600
  - Require Acme::Alien::DontPanic 2.5900 (Alien::Build 2.59)

2.5300    2022-08-12 14:30:46 -0600
  - Require Acme::Alien::DontPanic 2.5300 (Alien::Build 2.53)

2.5100    2022-08-03 10:01:16 -0600
  - Require Acme::Alien::DontPanic 2.5100 (Alien::Build 2.51)

2.2900    2020-08-12 12:08:04 -0600
  - Require Acme::Alien::DontPanic 2.2600 (Alien::Build 2.29)

2.1100    2020-03-09 03:13:09 -0600
  - Require Acme::Alien::DontPanic 2.1100 (Alien::Build 2.11)

2.0400    2020-02-05 19:12:07 -0700
  - Require Acme::Alien::DontPanic 2.0400 (Alien::Build 2.04)

2.0000    2020-02-02 06:31:10 -0700
  - Require Acme::Alien::DontPanic 2.0000 (Alien::Build 2.00)

1.9800    2020-01-30 08:11:32 -0700
  - Require Acme::Alien::DontPanic 1.9800 (Alien::Build 1.98)

1.96 2020-01-28 23:40:23 -0700
  - Require Acme::Alien::DontPanic 0.044 (Alien::Base 1.01 Alien::Base::ModuleBuild 0.044)

0.29 2017-07-30 06:38:55 -0400
  - Require Acme::Alien::DontPanic 0.042 (Alien::Base 0.75 Alien::Base::ModuleBuild 0.042)

0.28 Jan 31, 2017
  - Require Acme::Alien::DontPanic 0.030 (Alien::Base 0.030)

0.27  Nov 2, 2016
  - Require Acme::Alien::DontPanic 0.030 (Alien::Base 0.030)

0.26  Jun 22, 2016
  - Require Acme::Alien::DontPanic 0.029 (Alien::Base 0.029)

0.25  Mar 14, 2016
  - Require Acme::Alien::DontPanic 0.028 (Alien::Base 0.028)

0.24  Jan 20, 2016
  - Require Acme::Alien::DontPanic 0.027 (Alien::Base 0.025)

0.23  Jan 10, 2016
  - Require Acme::Alien::DontPanic 0.025 (Alien::Base 0.024)

0.22  Jul 20, 2015
  - Require Acme::Alien::DontPanic 0.025 (Alien::Base 0.022)

0.21  Jul 13, 2015
  - Require Acme::Alien::DontPanic 0.024 (Alien::Base 0.020)

0.20  Apr 27, 2015
  - Require Acme::Alien::DontPanic 0.023 (Alien::Base 0.017)

0.19  Apr 24, 2015
  - Require Acme::Alien::DontPanic 0.022 (Alien::Base 0.017)

0.18  Apr 22, 2015
  - Require Acme::Alien::DontPanic 0.021 (Alien::Base 0.016)

0.17  Mar 17, 2015
  - Require Acme::Alien::DontPanic 0.020 (Alien::Base 0.015)

0.16  Feb 25, 2015
  - Require Acme::Alien::DontPanic 0.018 (Alien::Base 0.014)

0.15  Feb 22, 2015
  - Fix bug in diagnostic test that would cause it to die
    if Acme::Alien::DontPanic was installed with install_type=system

0.14  Feb 22, 2015
  - Require Acme::Alien::DontPanic 0.016 (Alien::Base 0.012)

0.12  Feb 22, 2015
  - Require Acme::Alien::DontPanic 0.015 (Alien::Base 0.011)

0.11  Feb 17, 2015
 - Experimental work from 0.09 and 0.10 is now in Alien::Base
   removing from this distro

0.10  Feb 13, 2015
  - Use FFI::CheckLib now not experimental recursive (was _r)
    option.

0.09  Feb 12, 2015
  - Use FFI::CheckLib experimental _r option to see
    if we can more reliably test this via cpan testers.

0.08  Feb 6, 2015
  - Add README
  - Specify license in POD

0.07  Jan 22, 2015
  - FFI::Platypus 0.09 compatability

0.06  Jan 17, 2015
  - Migrate to FFI::Platypus
  - Non CPAN release

0.05  Jan 17, 2015
  - Require Acme::Alien::DontPanic 0.011

0.04  Oct 14, 2014
  - Require Acme::Alien::DontPanic 0.009

0.03  Oct 10, 2014
  - Add ddditional diagnostic for test phase

0.02  Oct 9, 2014
  - Require Acme::Alien::DontPanic 0.008
    which correctly builds dlls on MSWin32

0.01  Sep 11, 2014
  - Initial release


lib/Acme/Ford/Prefect/FFI.pm
package Acme::Ford::Prefect::FFI;

use strict;
use warnings;
use 5.008001;
use Acme::Alien::DontPanic ();
use FFI::Platypus::Declare;

# ABSTRACT: FFI test module for Alien::Base
# VERSION

=head1 SYNOPSIS

 use Acme::Ford::Prefect::FFI;
 
 my $answer = Acme::Ford::Prefect::FFI::answer(); # == 42 of course

=head1 DESCRIPTION

L<Alien::Base> comprises base classes to help in the construction of C<Alien::> modules.  Modules in the L<Alien> namespace are used to locate and install (if necessary)
external libraries needed by other Perl modules.

This module is a toy module to test the efficacy of the L<Alien::Base> system with its experimental FFI interfaces.  This module depends on another toy module 
L<Acme::Alien::DontPanic> which provides the needed libdontpanic library to be able to tell us the C<answer> to life, the universe and everything.

=head1 FUNCTIONS

=head2 answer

 my $answer = Acme::Ford::Prefect::FFI::answer();

Returns the answer to life the universe and everything.  Not exported.

=head1 SEE ALSO

=over 4

=item L<FFI::Platypus>

=item L<Alien::Base>

=item L<Alien>

=back

=cut

our($dll) = Acme::Alien::DontPanic->dynamic_libs;
die "no dll found for libdontpanic" unless $dll;
lib $dll;

attach answer => [] => int => '';

1;



Build.PL
BEGIN {
  unless(eval q{ use 5.008001; 1}) {
    print "Perl 5.008001 or better required\n";
    exit;
  }
}
use strict;
use warnings;

use Module::Build;
use Acme::Alien::DontPanic;

my $dontpanic = Acme::Alien::DontPanic->new;

my $cflags = $dontpanic->cflags;
my $libs   = $dontpanic->libs;

my $builder = Module::Build->new(
    module_name       => 'Acme::Ford::Prefect::FFI',
    dist_abstract     => 'FFI test for Alien::Base',
    dist_author       => 'Graham Ollis <plicease@wdlabs.com>',
    license           => 'perl',
    configure_requires => {
      'Module::Build' => 0.38,
      'Acme::Alien::DontPanic' => '0.030',
    },
    build_requires    => {
      'ExtUtils::CBuilder' => 0,
    },
    requires          => {
      'perl'                   => '5.008001',
      'FFI::Platypus'          => '0.08',
      'Acme::Alien::DontPanic' => '0.030',
      'FFI::CheckLib'          => '0.11',
    },
    meta_merge => {
      resources  => {
        repository => "http://github.com/Perl5-Alien/Acme-Ford-Prefect-FFI",
        bugtracker => "http://github.com/Perl5-Alien/Acme-Ford-Prefect-FFI/issues",
        x_MailingList => 'https://groups.google.com/forum/#!forum/perl5-alien',
        x_IRC => "irc://irc.perl.org/#native",
      },
    },  
);

$builder->add_to_cleanup('Acme-Ford-Prefect-FFI*');

$builder->create_build_script;

